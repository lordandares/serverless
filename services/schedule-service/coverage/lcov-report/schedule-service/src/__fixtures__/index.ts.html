<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for schedule-service/src/__fixtures__/index.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">schedule-service/src/__fixtures__</a> index.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>4/4</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>4/4</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145</td><td class="line-coverage quiet"><span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">export const scheduleDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    areas: ['area1'],
    enabled: true,
    locations: ['location1'],
    name: 'Schedule One',
    included: {},
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
    strategy: {
      options: {
        duration: {
          unit: 'minute',
          value: 15,
        },
        frequency: {
          unit: 'minute',
          value: 15,
        },
      },
      type: 'stopwatch',
    },
    type: 'visit',
  },
  endAt: '2019-10-1500:00:00.000Z',
  groupType: 'schedule',
  pk: 'application1-schedule',
  scheduleId: 'schedule1',
  sk: '2019-10-13T00:00:00.000Z-schedule',
  startAt: '2019-10-1400:00:00.000Z',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}
&nbsp;
export const schedulePayload = {
  areas: ['area1'],
  enabled: true,
  endAt: '2019-10-1500:00:00.000Z',
  locations: ['location1'],
  name: 'Schedule One',
  serviceHours: {
    hours: [
      {
        description: 'MON 11:00 - TUE 00:00',
        end: 86400000,
        start: 39600000,
        type: 'DEFAULT',
      },
    ],
    timezone: 'Australia/Melbourne',
  },
  startAt: '2019-10-1400:00:00.000Z',
  strategy: {
    options: {
      duration: {
        unit: 'minute',
        value: 15,
      },
      frequency: {
        unit: 'minute',
        value: 15,
      },
    },
    type: 'stopwatch',
  },
  type: 'visit',
}
&nbsp;
export const occurrenceDocument = {
  applicationId: 'application1',
  createdAt: '2019-01-01T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  endAt: '2019-02-02T00:00:00.000Z',
  groupType: 'occurrence',
  location: 'location1',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-01-01T00:00-occurrence',
  startAt: '2019-02-01T00:00:00.000Z',
  type: 'task',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}
&nbsp;
export const areaDocument = {
  _id: 'location1',
  application: '56789671d4c6488b128c7f9c',
  name: 'WRNMMC Facility',
  altName: 'WRNMMC Facility',
  type: 'location',
  tags: [],
  address: {
    street: '8901 Rockville Pike',
    street2: '',
    city: 'Bethesda',
    state: 'MD',
    postal: '20889',
    country: 'US',
  },
  keywords: ['hospital'],
  areaSize: 536559.55514508,
  serviceHours: {
    hours: [
      {
        description: 'MON 00:00 - TUE 00:00',
        duration: 86400000,
        end: 86400000,
        start: 0,
        type: 'DEFAULT',
      },
    ],
    timezone: 'America/New_York',
  },
  deleted: false,
  timezone: 'America/New_York',
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Nov 25 2019 18:46:45 GMT+0000 (Coordinated Universal Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
