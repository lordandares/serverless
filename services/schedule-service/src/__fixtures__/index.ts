import { Context } from 'aws-lambda'

export const areaDocument = {
  _id: 'location1',
  address: {
    city: 'Bethesda',
    country: 'US',
    postal: '20889',
    state: 'MD',
    street: '8901 Rockville Pike',
    street2: '',
  },
  altName: 'WRNMMC Facility',
  application: '56789671d4c6488b128c7f9c',
  areaSize: 536559.55514508,
  deleted: false,
  floorsRef: [0],
  groupType: 'location',
  keywords: ['hospital'],
  name: 'WRNMMC Facility',
  serviceHours: {
    hours: [
      {
        description: 'MON 00:00 - TUE 00:00',
        duration: 86400000,
        end: 86400000,
        start: 0,
        type: 'DEFAULT',
      },
    ],
    timezone: 'America/New_York',
  },
  tags: [],
  timezone: 'America/New_York',
}

export const occurrenceDocument = {
  applicationId: 'application1',
  createdAt: '2019-01-01T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    occurrenceInterval: [1, 2],
    scheduleName: 'Schedule 1',
    serviceInterval: [1, 2],
    timezone: 'Australia/Melbourne',
    type: 'visit',
  },
  endAt: '2019-02-01T23:59:59.999Z',
  groupType: 'occurrence',
  locationId: '5d163137b8b3b7000127edd1',
  location_endAt_occurrenceId: 'location1-2019-02-02T00:00:00.000Z-occurrence1',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-01-01T00:00-occurrence1',
  startAt: '2019-02-01T00:00:00.000Z',
  status: 'pending',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}

export const occurrenceTimerDocument = {
  pk: `timer-2019-01-01T00:00`,
  sk: `timer#application1-occurrence#2019-01-01T00:00-occurrence`,
  endAt: '2019-01-01T23:59:59.999Z',
  groupType: `occurrence-timer`,
  scheduleId: 'schedule1',
  targetArn: 'someArn',
}

export const scheduleDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    areas: ['area1'],
    enabled: true,
    included: {},
    locations: ['location1'],
    name: 'Schedule One',
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
    strategy: {
      options: {
        duration: {
          unit: 'minute',
          value: 15,
        },
        frequency: {
          unit: 'minute',
          value: 15,
        },
      },
      type: 'stopwatch',
    },
    type: 'visit',
  },
  endAt: '2019-10-1500:00:00.000Z',
  groupType: 'schedule',
  pk: 'application1-schedule',
  scheduleId: 'schedule1',
  sk: '2019-10-13T00:00:00.000Z-schedule1',
  startAt: '2019-10-1400:00:00.000Z',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}

export const scheduleLocationDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: '123456789',
    label: 'label',
    type: 'system',
  },
  data: {
    name: 'Schedule One',
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
  },
  groupType: 'location',
  pk: 'application1-location',
  schedules: ['schedule1'],
  sk: 'location1',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: '123456789',
    label: 'label',
    type: 'system',
  },
}

export const schedulePayload = {
  areas: ['area1'],
  enabled: true,
  endAt: '2019-10-1500:00:00.000Z',
  locations: ['location1'],
  name: 'Schedule One',
  serviceHours: {
    hours: [
      {
        description: 'MON 11:00 - TUE 00:00',
        end: 86400000,
        start: 39600000,
        type: 'DEFAULT',
      },
    ],
    timezone: 'Australia/Melbourne',
  },
  startAt: '2019-10-1400:00:00.000Z',
  strategy: {
    options: {
      duration: {
        unit: 'minute',
        value: 15,
      },
      frequency: {
        unit: 'minute',
        value: 15,
      },
    },
    type: 'stopwatch',
  },
  type: 'visit',
}

export const rulePayload = {
  applicationId: 'application-id-1111',
  data: {
    occurrenceInterval: [1580761454998, 1580762054997],
  },
  hook: {
    endpoint: 'sns:endpoint:uri',
    type: 'sns',
  },
  locationId: 'location-id-11111',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-06-02T00:00:00.000Z-occurrence1',
  startAt: '2019-02-01T00:00:00.000Z',
  type: 'visit',
}

export const rulePatternDocument = {
  createdAt: '2019-06-02T00:00:00.000Z',
  createdBy: {
    id: 'created-by-id-11111',
    label: 'created-by-label-1111',
    type: 'created-by-type-11111',
  },
  groupType: 'rule-pattern',
  matches: {
    occurrence1: {
      pk: 'application1-occurrence',
      sk: '2019-06-01T00:00:00.000Z-occurrence1',
    },
  },
  pk: 'rule-pattern-application-id-11111-location-id-11111',
  sk: 'visit',
  updatedAt: '2019-06-02T00:00:00.000Z',
}

export const context: Context = {
  awsRequestId: 'aws-request-id',
  callbackWaitsForEmptyEventLoop: false,
  clientContext: undefined,
  done: jest.fn(),
  fail: jest.fn(),
  functionName: 'function-name',
  functionVersion: 'function-version',
  getRemainingTimeInMillis: jest.fn(),
  identity: undefined,
  invokedFunctionArn: 'invoked-function-arn',
  logGroupName: 'log-group-name',
  logStreamName: 'log-stream-name',
  memoryLimitInMB: 0,
  succeed: jest.fn(),
}
