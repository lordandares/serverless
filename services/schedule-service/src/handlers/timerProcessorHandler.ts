import { errors, schemas } from '@lighthouse/serverless-common'
import { DynamoDB, SNS } from 'aws-sdk'
import { DocumentClient } from 'aws-sdk/clients/dynamodb'
import * as BPromise from 'bluebird'
import * as moment from 'moment'

declare var process: {
  env: {
    TABLE_SCHEDULES: string
  }
}

interface Document {
  pk: string
  sk: string
}

interface TimerError {
  error: Error
  timer: schemas.TimerDocumentSchema
}

export async function timerProcessorHandler(): Promise<
  schemas.TimerDocumentSchema[]
> {
  try {
    const documentClient = new DynamoDB.DocumentClient()
    const sns = new SNS()
    const tableName = process.env.TABLE_SCHEDULES

    const mNow = moment.utc()

    // NOTE: we query both minutes in case when processing we go into the following minute
    const currentMinute = mNow.format('YYYY-MM-DDTHH:mm')
    const previousMinute = mNow.subtract(1, 'minute').format('YYYY-MM-DDTHH:mm')

    const currentMinuteQueryParams: DocumentClient.QueryInput = {
      ExpressionAttributeNames: {
        '#pk': 'pk',
      },
      ExpressionAttributeValues: {
        ':pk': `timer-${currentMinute}`,
      },
      KeyConditionExpression: '#pk = :pk',
      ScanIndexForward: false,
      TableName: tableName,
    }

    const previousMinuteQueryParams: DocumentClient.QueryInput = {
      ExpressionAttributeNames: {
        '#pk': 'pk',
      },
      ExpressionAttributeValues: {
        ':pk': `timer-${previousMinute}`,
      },
      KeyConditionExpression: '#pk = :pk',
      ScanIndexForward: false,
      TableName: tableName,
    }

    console.info(
      `TimerProcessorHandler: querying for current minute timers with ${currentMinute}`,
    )
    console.info(
      `TimerProcessorHandler: querying for previous minute timers with ${previousMinute}`,
    )

    const currentMinuteQueryResponse = await documentClient
      .query(currentMinuteQueryParams)
      .promise()
    const previousMinuteQueryResponse = await documentClient
      .query(previousMinuteQueryParams)
      .promise()

    const { Items: currentMinuteItems } = currentMinuteQueryResponse
    const { Items: previousMinuteItems } = previousMinuteQueryResponse

    const expiredTimers =
      currentMinuteItems && previousMinuteItems
        ? [...currentMinuteItems, ...previousMinuteItems]
        : []

    if (!expiredTimers.length) {
      console.info('TimerProcessorHandler: no timers to process')
      return []
    }

    console.info(
      `TimerProcessorHandler: processing ${expiredTimers.length} timers`,
    )

    const deleteTimerErrors: TimerError[] = []
    const processTimerErrors: TimerError[] = []

    const processedTimers = await BPromise.reduce(
      expiredTimers,
      async (accum: Document[], expiredTimer: schemas.TimerDocumentSchema) => {
        const { pk: timerPk, sk: timerSk, targetArn } = expiredTimer

        try {
          const parts = timerSk.split('#')
          const targetPk: string = parts[1]
          const targetSk: string = parts[2]

          const snsParams: SNS.Types.PublishInput = {
            Message: JSON.stringify({
              pk: targetPk,
              sk: targetSk,
            }),
            TargetArn: targetArn,
          }

          console.info(
            `TimerProcessorHandler: publishing to targetArn: ${targetArn}`,
          )

          await sns.publish(snsParams).promise()

          accum.push({ pk: timerPk, sk: timerSk })

          return accum
        } catch (error) {
          console.warn(
            `TimerProcessorHandler: publishing to targetArn: ${targetArn} failed`,
          )
          processTimerErrors.push({ error, timer: expiredTimer })
          return accum
        }
      },
      [],
    )

    console.info(
      `TimerProcessorHandler: processed ${processedTimers.length} timers`,
    )

    if (processedTimers.length) {
      console.info(
        `TimerProcessorHandler: removing ${processedTimers.length} timers`,
      )

      await BPromise.each(processedTimers, async timer => {
        try {
          const deleteParams = {
            Key: { pk: timer.pk, sk: timer.sk },
            TableName: tableName,
          }
          await documentClient.delete(deleteParams).promise()
        } catch (error) {
          console.warn('TimerProcessorHandler: deleting timer failed')
          deleteTimerErrors.push({ error, timer })
        }
      })
    }

    if (processTimerErrors.length) {
      const message = `TimerProcessorHandler: caught ${
        processTimerErrors.length
      } processing errors for ${expiredTimers.length} timers`

      throw new errors.ApplicationError({
        data: { errors: JSON.stringify(processTimerErrors) },
        message,
      })
    }

    if (deleteTimerErrors.length) {
      const message = `TimerProcessorHandler: caught ${
        deleteTimerErrors.length
      } delete errors for ${deleteTimerErrors.length} timers`

      console.error(message)

      throw new errors.ApplicationError({
        data: { errors: JSON.stringify(deleteTimerErrors) },
        message,
      })
    }

    return processedTimers
  } catch (error) {
    console.error('TimerProcessorHandlerError', {
      error,
    })

    throw error
  }
}
