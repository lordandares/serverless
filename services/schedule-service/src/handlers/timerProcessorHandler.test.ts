import * as AWS from 'aws-sdk-mock'
import * as MockDate from 'mockdate'

import { occurrenceTimerDocument } from '../__fixtures__'

import { timerProcessorHandler } from './timerProcessorHandler'

beforeEach(() => {
  process.env.IS_OFFLINE = 'true'
  process.env.TABLE_SCHEDULES = 'table-schedules'
})

afterEach(() => {
  AWS.restore()
})

test('successful timer processing', async () => {
  expect.assertions(1)

  const deleteSpy = jest.fn().mockResolvedValue({})

  const querySpy = jest
    .fn()
    .mockResolvedValueOnce({ Items: [{ ...occurrenceTimerDocument }] })
    .mockResolvedValueOnce({
      Items: [
        {
          ...occurrenceTimerDocument,
          pk: 'timer-2018-12-31T23:59',
          sk: 'timer#application1-occurrence#timer-2018-12-31T23:59-occurrence',
        },
      ],
    })

  const snsSpy = jest.fn().mockResolvedValue('success')

  AWS.mock('DynamoDB.DocumentClient', 'delete', deleteSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)
  AWS.mock('SNS', 'publish', snsSpy)

  const result = await timerProcessorHandler()

  expect(result).toMatchInlineSnapshot(`
Array [
  Object {
    "pk": "timer-2019-01-01T00:00",
    "sk": "timer#application1-occurrence#2019-01-01T00:00-occurrence",
  },
  Object {
    "pk": "timer-2018-12-31T23:59",
    "sk": "timer#application1-occurrence#timer-2018-12-31T23:59-occurrence",
  },
]
`)
})

test('error when querying for timers', async () => {
  expect.assertions(1)

  const querySpy = jest.fn().mockRejectedValue(new Error('NetworkError'))

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  try {
    await timerProcessorHandler()
  } catch (error) {
    expect(error).toMatchInlineSnapshot(`[Error: NetworkError]`)
  }
})

test('handles no timers to process', async () => {
  expect.assertions(1)

  const querySpy = jest
    .fn()
    .mockResolvedValueOnce({ Items: [] })
    .mockResolvedValueOnce({ Items: [] })

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const result = await timerProcessorHandler()

  expect(result).toMatchInlineSnapshot(`Array []`)
})

test('handles no timers to process when missing items', async () => {
  expect.assertions(1)

  const querySpy = jest
    .fn()
    .mockResolvedValueOnce({ Items: null })
    .mockResolvedValueOnce({ Items: null })

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const result = await timerProcessorHandler()

  expect(result).toMatchInlineSnapshot(`Array []`)
})

test('handles sns error but continues to process and throws after', async () => {
  expect.assertions(1)

  const deleteSpy = jest.fn().mockResolvedValue({})

  const querySpy = jest
    .fn()
    .mockResolvedValueOnce({
      Items: [{ ...occurrenceTimerDocument }, { ...occurrenceTimerDocument }],
    })
    .mockResolvedValueOnce({ Items: [] })

  const snsSpy = jest
    .fn()
    .mockRejectedValueOnce(new Error('Processing Error'))
    .mockResolvedValueOnce('success')

  AWS.mock('DynamoDB.DocumentClient', 'delete', deleteSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)
  AWS.mock('SNS', 'publish', snsSpy)

  try {
    await timerProcessorHandler()
  } catch (error) {
    expect(error).toMatchSnapshot()
  }
})

test('handles delete error but continues to process and throws after', async () => {
  expect.assertions(1)

  const deleteSpy = jest
    .fn()
    .mockResolvedValueOnce({})
    .mockRejectedValueOnce(new Error('Delete Error'))

  const querySpy = jest
    .fn()
    .mockResolvedValueOnce({
      Items: [{ ...occurrenceTimerDocument }, { ...occurrenceTimerDocument }],
    })
    .mockResolvedValueOnce({ Items: [] })

  const snsSpy = jest
    .fn()
    .mockResolvedValueOnce('success')
    .mockResolvedValueOnce('success')

  AWS.mock('DynamoDB.DocumentClient', 'delete', deleteSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)
  AWS.mock('SNS', 'publish', snsSpy)

  try {
    await timerProcessorHandler()
  } catch (error) {
    expect(error).toMatchSnapshot()
  }
})

test('skips removing timers when none processed', async () => {
  expect.assertions(1)

  const deleteSpy = jest.fn().mockResolvedValue({})

  const querySpy = jest
    .fn()
    .mockResolvedValue({
      Items: [{ ...occurrenceTimerDocument }, { ...occurrenceTimerDocument }],
    })
    .mockResolvedValueOnce({ Items: [] })

  const snsSpy = jest
    .fn()
    .mockRejectedValueOnce(new Error('Processing Error 1'))
    .mockRejectedValueOnce(new Error('Processing Error 2'))

  AWS.mock('DynamoDB.DocumentClient', 'delete', deleteSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)
  AWS.mock('SNS', 'publish', snsSpy)

  try {
    await timerProcessorHandler()
  } catch (error) {
    expect(deleteSpy).toHaveBeenCalledTimes(0)
  }
})
