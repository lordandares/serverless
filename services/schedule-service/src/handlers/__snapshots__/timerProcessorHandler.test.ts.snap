// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles delete error but continues to process and throws after 1`] = `[ApplicationError: TimerProcessorHandler: caught 1 delete errors for 1 timers]`;

exports[`handles sns error but continues to process and throws after 1`] = `[ApplicationError: TimerProcessorHandler: caught 1 processing errors for 2 timers]`;
